File: hadrian/src/test/scala/lib1/rand.scala
// Copyright (C) 2014  Open Data ("Open Data" refers to
// one or more of the following companies: Open Data Partners LLC,
// Open Data Research LLC, or Open Data Capital LLC.)
// 
// This file is part of Hadrian.
// 
// Licensed under the Hadrian Personal Use and Evaluation License (PUEL);
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://raw.githubusercontent.com/opendatagroup/hadrian/master/LICENSE
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package test.scala.lib1.rand

import scala.collection.JavaConversions._

import org.junit.runner.RunWith

import org.scalatest.FlatSpec
import org.scalatest.junit.JUnitRunner
import org.scalatest.Matchers

import com.opendatagroup.hadrian.data._
import com.opendatagroup.hadrian.jvmcompiler._
import com.opendatagroup.hadrian.errors._
import test.scala._

@RunWith(classOf[JUnitRunner])
class Lib1RandSuite extends FlatSpec with Matchers {
  "numerical rand engine" must "generate random integers" taggedAs(Lib1, Lib1Rand) in {
    val engine1 = PFAEngine.fromYaml("""
input: "null"
output: int
randseed: 12345
action: {rand.int: []}
""").head
    engine1.action(null) should be (1553932502)
    engine1.action(null) should be (-2090749135)
    engine1.action(null) should be (-287790814)

    val engine2 = PFAEngine.fromYaml("""
input: "null"
output: int
randseed: 12345
action: {rand.int: [5, 10]}
""").head
    engine2.action(null) should be (6)
    engine2.action(null) should be (5)
    engine2.action(null) should be (6)
  }

  it must "generate random longs" taggedAs(Lib1, Lib1Rand) in {
    val engine1 = PFAEngine.fromYaml("""
input: "null"
output: long
randseed: 12345
action: {rand.long: []}
""").head
    engine1.action(null) should be (6674089274190705457L)
    engine1.action(null) should be (-1236052134575208584L)
    engine1.action(null) should be (-3078921119283744887L)

    val engine2 = PFAEngine.fromYaml("""
input: "null"
output: long
randseed: 12345
action: {rand.long: [5, 10]}
""").head
    engine2.action(null) should be (7)
    engine2.action(null) should be (9)
    engine2.action(null) should be (7)
  }

  it must "generate random floats" taggedAs(Lib1, Lib1Rand) in {
    val engine = PFAEngine.fromYaml("""
input: "null"
output: float
randseed: 12345
action: {rand.float: [5, 10]}
""").head
    engine.action(null).asInstanceOf[java.lang.Number].doubleValue should be (6.8090153 +- 0.00001)
    engine.action(null).asInstanceOf[java.lang.Number].doubleValue should be (7.5660477 +- 0.00001)
    engine.action(null).asInstanceOf[java.lang.Number].doubleValue should be (9.664968 +- 0.00001)
  }

  it must "generate random doubles" taggedAs(Lib1, Lib1Rand) in {
    val engine = PFAEngine.fromYaml("""
input: "null"
output: double
randseed: 12345
action: {rand.double: [5, 10]}
""").head
    engine.action(null).asInstanceOf[java.lang.Number].doubleValue should be (6.809015535802359 +- 0.00001)
    engine.action(null).asInstanceOf[java.lang.Number].doubleValue should be (9.664967426442704 +- 0.00001)
    engine.action(null).asInstanceOf[java.lang.Number].doubleValue should be (9.165456744855119 +- 0.00001)
  }

  "categorical rand engine" must "make a random choice" taggedAs(Lib1, Lib1Rand) in {
    val engine = PFAEngine.fromYaml("""
input:
  type: array
  items: string
output: string
randseed: 12345
action: {rand.choice: input}
""").head
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))) should be ("two")
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))) should be ("one")
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))) should be ("two")
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))) should be ("four")
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))) should be ("one")
  }

  it must "make random choices with replacement" taggedAs(Lib1, Lib1Rand) in {
    val engine = PFAEngine.fromYaml("""
input:
  type: array
  items: string
output:
  type: array
  items: string
randseed: 12345
action: {rand.choices: [3, input]}
""").head
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("two", "one", "two"))
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("four", "one", "five"))
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("one", "three", "two"))
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("five", "three", "three"))
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("one", "two", "three"))
  }

  it must "make random samples without replacement" taggedAs(Lib1, Lib1Rand) in {
    val engine = PFAEngine.fromYaml("""
input:
  type: array
  items: string
output:
  type: array
  items: string
randseed: 12345
action: {rand.sample: [3, input]}
""").head
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("one", "two", "four"))
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("one", "three", "five"))
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("two", "three", "five"))
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("one", "two", "three"))
    engine.action(PFAArray.fromVector(Vector("one", "two", "three", "four", "five"))).asInstanceOf[PFAArray[String]].toVector should be (Vector("three", "four", "five"))
  }

  it must "generate random strings" taggedAs(Lib1, Lib1Rand) in {
    val engine1 = PFAEngine.fromYaml("""
input: "null"
output: string
randseed: 12345
action: {rand.string: [10]}
""").head
    engine1.action(null) should be ("㽏彴爜뛞謪ᕼ㻇鬜䣚")
    engine1.action(null) should be ("瓔ꣃ잘젹랴⭺篦杌⤪晀")
    engine1.action(null) should be ("鈩睧ᤜ꺊뎇頀鴢豏荴薱")

    val engine2 = PFAEngine.fromYaml("""
input: "null"
output: string
randseed: 12345
action: {rand.string: [10, {string: "abcdefghijklmnopqrstuvwxyz0123456789"}]}
""").head
    engine2.action(null) should be ("5e7a5ehsz7")
    engine2.action(null) should be ("fg886qlti1")
    engine2.action(null) should be ("352quv3x91")

    val engine3 = PFAEngine.fromYaml("""
input: "null"
output: string
randseed: 12345
action: {rand.string: [10, 33, 127]}
""").head
    engine3.action(null) should be ("HaRmFM:wT*")
    engine3.action(null) should be ("TO9!5i68Q4")
    engine3.action(null) should be ("V\"W/%>46<z")
  }

  it must "generate random bytes" taggedAs(Lib1, Lib1Rand) in {
    val engine1 = PFAEngine.fromYaml("""
input: "null"
output: bytes
randseed: 12345
action: {rand.bytes: [10]}
""").head
    engine1.action(null).asInstanceOf[Array[Byte]].toVector should be (Vector(-42, 32, -97, 92, 49, -77, 97, -125, 34, -87))
    engine1.action(null).asInstanceOf[Array[Byte]].toVector should be (Vector(120, 7, -56, -22, -114, 121, 69, -43, -119, -17))
    engine1.action(null).asInstanceOf[Array[Byte]].toVector should be (Vector(127, -22, -109, 83, 100, 64, -22, 31, -101, 73))

    val engine2 = PFAEngine.fromYaml("""
input: "null"
output: bytes
randseed: 12345
action: {rand.bytes: [10, {base64: "YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXowMTIzNDU2Nzg5"}]}
""").head
    engine2.action(null).asInstanceOf[Array[Byte]].toVector should be (Vector(53, 101, 55, 97, 53, 101, 104, 115, 122, 55))
    engine2.action(null).asInstanceOf[Array[Byte]].toVector should be (Vector(102, 103, 56, 56, 54, 113, 108, 116, 105, 49))
    engine2.action(null).asInstanceOf[Array[Byte]].toVector should be (Vector(51, 53, 50, 113, 117, 118, 51, 120, 57, 49))

    val engine3 = PFAEngine.fromYaml("""
input: "null"
output: bytes
randseed: 12345
action: {rand.bytes: [10, 33, 127]}
""").head
    engine3.action(null).asInstanceOf[Array[Byte]].toVector should be (Vector(72, 97, 82, 109, 70, 77, 58, 119, 84, 42))
    engine3.action(null).asInstanceOf[Array[Byte]].toVector should be (Vector(84, 79, 57, 33, 53, 105, 54, 56, 81, 52))
    engine3.action(null).asInstanceOf[Array[Byte]].toVector should be (Vector(86, 34, 87, 47, 37, 62, 52, 54, 60, 122))
  }

  it must "generate random UUIDs" taggedAs(Lib1, Lib1Rand) in {
    val engine1 = PFAEngine.fromYaml("""
input: "null"
output: string
randseed: 12345
action: {rand.uuid: []}
""").head
    engine1.action(null) should be ("5c9f20d5-8361-4331-8ed8-a921eac80778")
    engine1.action(null) should be ("d545798e-09a4-4f89-8393-ea7f1fea4064")
    engine1.action(null) should be ("3c4b499b-090b-46fa-895f-9fb9fcf0e5bc")

    val engine2 = PFAEngine.fromYaml("""
input: "null"
output: string
action: {s.substr: [{rand.uuid: []}, 14, 15]}
""").head
    for (i <- 0 until 1000)
      engine2.action(null) should be ("4")

    val engine3 = PFAEngine.fromYaml("""
input: "null"
output: string
action: {s.substr: [{rand.uuid: []}, 19, 20]}
""").head
    for (i <- 0 until 1000)
      engine3.action(null) should be ("8")
  }

  "random distributions" must "generate Gaussian deviates" taggedAs(Lib1, Lib1Rand) in {
    val engine = PFAEngine.fromYaml("""
input: "null"
output: double
randseed: 12345
action: {rand.gaussian: [10, 2]}
""").head
    engine.action(null).asInstanceOf[java.lang.Number].doubleValue should be (9.624382020682177 +- 0.00001)
    engine.action(null).asInstanceOf[java.lang.Number].doubleValue should be (11.176872610230959 +- 0.00001)
    engine.action(null).asInstanceOf[java.lang.Number].doubleValue should be (11.897609560880085 +- 0.00001)
  }

}

File: titus/titus/lib/rand.py
#!/usr/bin/env python

# Copyright (C) 2014  Open Data ("Open Data" refers to
# one or more of the following companies: Open Data Partners LLC,
# Open Data Research LLC, or Open Data Capital LLC.)
# 
# This file is part of Hadrian.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

from titus.fcn import Fcn
from titus.fcn import LibFcn
from titus.signature import Sig
from titus.signature import Sigs
from titus.signature import Lifespan
from titus.signature import PFAVersion
from titus.datatype import *
from titus.errors import *
from titus.lib.core import INT_MIN_VALUE, INT_MAX_VALUE, LONG_MIN_VALUE, LONG_MAX_VALUE
import titus.P as P

provides = {}
def provide(fcn):
    provides[fcn.name] = fcn

prefix = "rand."

########################################################### raw numbers of various types

class RandomInt(LibFcn):
    name = prefix + "int"
    sig = Sigs([Sig([], P.Int()),
                Sig([{"low": P.Int()}, {"high": P.Int()}], P.Int())])
    errcodeBase = 34000
    def __call__(self, state, scope, pos, paramTypes, *args):
        if len(args) == 0:
            return state.rand.randint(INT_MIN_VALUE, INT_MAX_VALUE)
        else:
            if args[1] <= args[0]: raise PFARuntimeException("high must be greater than low", self.errcodeBase + 0, self.name, pos)
            return state.rand.randint(args[0], args[1] - 1)  # Python's randint has an inclusive upper bound
provide(RandomInt())

class RandomLong(LibFcn):
    name = prefix + "long"
    sig = Sigs([Sig([], P.Long()),
                Sig([{"long": P.Long()}, {"high": P.Long()}], P.Long())])
    errcodeBase = 34010
    def __call__(self, state, scope, pos, paramTypes, *args):
        if len(args) == 0:
            return state.rand.randint(LONG_MIN_VALUE, LONG_MAX_VALUE)
        else:
            if args[1] <= args[0]: raise PFARuntimeException("high must be greater than low", self.errcodeBase + 0, self.name, pos)
            return state.rand.randint(args[0], args[1] - 1)  # Python's randint has an inclusive upper bound
provide(RandomLong())

class RandomFloat(LibFcn):
    name = prefix + "float"
    sig = Sig([{"low": P.Float()}, {"high": P.Float()}], P.Float())
    errcodeBase = 34020
    def __call__(self, state, scope, pos, paramTypes, low, high):
        if high <= low: raise PFARuntimeException("high must be greater than low", self.errcodeBase + 0, self.name, pos)
        return state.rand.uniform(low, high)
provide(RandomFloat())

class RandomDouble(LibFcn):
    name = prefix + "double"
    sig = Sig([{"low": P.Double()}, {"high": P.Double()}], P.Double())
    errcodeBase = 34030
    def __call__(self, state, scope, pos, paramTypes, low, high):
        if high <= low: raise PFARuntimeException("high must be greater than low", self.errcodeBase + 0, self.name, pos)
        return state.rand.uniform(low, high)
provide(RandomDouble())

########################################################### items from arrays

class RandomChoice(LibFcn):
    name = prefix + "choice"
    sig = Sig([{"population": P.Array(P.Wildcard("A"))}], P.Wildcard("A"))
    errcodeBase = 34040
    def __call__(self, state, scope, pos, paramTypes, population):
        if len(population) == 0:
            raise PFARuntimeException("population must not be empty", self.errcodeBase + 0, self.name, pos)
        return population[state.rand.randint(0, len(population) - 1)]
provide(RandomChoice())

class RandomChoices(LibFcn):
    name = prefix + "choices"
    sig = Sig([{"size": P.Int()}, {"population": P.Array(P.Wildcard("A"))}], P.Array(P.Wildcard("A")))
    errcodeBase = 34050
    def __call__(self, state, scope, pos, paramTypes, size, population):
        if len(population) == 0:
            raise PFARuntimeException("population must not be empty", self.errcodeBase + 0, self.name, pos)
        return [population[state.rand.randint(0, len(population) - 1)] for x in xrange(size)]
provide(RandomChoices())

class RandomSample(LibFcn):
    name = prefix + "sample"
    sig = Sig([{"size": P.Int()}, {"population": P.Array(P.Wildcard("A"))}], P.Array(P.Wildcard("A")))
    errcodeBase = 34060
    def __call__(self, state, scope, pos, paramTypes, size, population):
        if len(population) == 0:
            raise PFARuntimeException("population must not be empty", self.errcodeBase + 0, self.name, pos)
        if len(population) < size:
            raise PFARuntimeException("population smaller than requested subsample", self.errcodeBase + 1, self.name, pos)
        return state.rand.sample(population, size)
provide(RandomSample())
        
########################################################### deviates from a histogram

class RandomHistogram(LibFcn):
    name = prefix + "histogram"
    sig = Sigs([Sig([{"distribution": P.Array(P.Double())}], P.Int()),
                Sig([{"distribution": P.Array(P.WildRecord("A", {"prob": P.Double()}))}], P.Wildcard("A"))])
    errcodeBase = 34070
    def selectIndex(self, rand, distribution, pos):
        cumulativeSum = [0.0]
        for x in distribution:
            if math.isnan(x) or math.isinf(x):
                raise PFARuntimeException("distribution must be finite", self.errcodeBase + 1, self.name, pos)
            elif x < 0:
                raise PFARuntimeException("distribution must be non-negative", self.errcodeBase + 2, self.name, pos)
            cumulativeSum.append(cumulativeSum[-1] + x)
        total = cumulativeSum[-1]
        if total == 0.0:
            raise PFARuntimeException("distribution must be non-empty", self.errcodeBase + 0, self.name, pos)
        position = rand.uniform(0.0, total)
        for i, y in enumerate(cumulativeSum):
            if position < y:
                return i - 1
    def __call__(self, state, scope, pos, paramTypes, distribution):
        if isinstance(paramTypes[-1], dict) and paramTypes[-1].get("type") == "record":
            probs = [x["prob"] for x in distribution]
            index = self.selectIndex(state.rand, probs, pos)
            return distribution[index]
        else:
            return self.selectIndex(state.rand, distribution, pos)
provide(RandomHistogram())

########################################################### strings and byte arrays

class RandomString(LibFcn):
    name = prefix + "string"
    sig = Sigs([Sig([{"size": P.Int()}], P.String()),
                Sig([{"size": P.Int()}, {"population": P.String()}], P.String()),
                Sig([{"size": P.Int()}, {"low": P.Int()}, {"high": P.Int()}], P.String())])
    errcodeBase = 34080
    def __call__(self, state, scope, pos, paramTypes, size, *args):
        if size <= 0: raise PFARuntimeException("size must be positive", self.errcodeBase + 0, self.name, pos)
        if len(args) == 0:
            return "".join(unichr(state.rand.randint(1, 0xD800)) for x in xrange(size))
        elif len(args) == 1:
            if len(args[0]) == 0:
                raise PFARuntimeException("population must be non-empty", self.errcodeBase + 3, self.name, pos)
            return "".join(args[0][state.rand.randint(0, len(args[0]) - 1)] for x in xrange(size))
        else:
            low, high = args
            if high <= low: raise PFARuntimeException("high must be greater than low", self.errcodeBase + 1, self.name, pos)
            if low < 1 or low > 0xD800 or high < 1 or high > 0xD800: raise PFARuntimeException("invalid char", self.errcodeBase + 2, self.name, pos)
            return u"".join(unichr(state.rand.randint(low, high - 1)) for x in xrange(size))
provide(RandomString())

class RandomBytes(LibFcn):
    name = prefix + "bytes"
    sig = Sigs([Sig([{"size": P.Int()}], P.Bytes()),
                Sig([{"size": P.Int()}, {"population": P.Bytes()}], P.Bytes()),
                Sig([{"size": P.Int()}, {"low": P.Int()}, {"high": P.Int()}], P.Bytes())])
    errcodeBase = 34090
    def __call__(self, state, scope, pos, paramTypes, size, *args):
        if size <= 0: raise PFARuntimeException("size must be positive", self.errcodeBase + 0, self.name, pos)
        if len(args) == 0:
            return "".join(chr(state.rand.randint(0, 255)) for x in xrange(size))
        elif len(args) == 1:
            if len(args[0]) == 0:
                raise PFARuntimeException("population must be non-empty", self.errcodeBase + 3, self.name, pos)
            return "".join(args[0][state.rand.randint(0, len(args[0]) - 1)] for x in xrange(size))
        else:
            low, high = args
            if high <= low: raise PFARuntimeException("high must be greater than low", self.errcodeBase + 1, self.name, pos)
            if low < 0 or low > 255 or high < 0 or high > 256: raise PFARuntimeException("invalid byte", self.errcodeBase + 2, self.name, pos)
            return "".join(chr(state.rand.randint(low, high - 1)) for x in xrange(size))
provide(RandomBytes())

class RandomUUID(LibFcn):
    name = prefix + "uuid"
    sig = Sig([], P.String(), Lifespan(None, PFAVersion(0, 7, 2), PFAVersion(0, 9, 0), "use rand.uuid4 instead"))
    errcodeBase = 34100
    def __call__(self, state, scope, pos, paramTypes):
        return "{0:08x}-{1:04x}-4{2:03x}-8{3:03x}-{4:016x}".format(state.rand.getrandbits(32), state.rand.getrandbits(16), state.rand.getrandbits(12), state.rand.getrandbits(12), state.rand.getrandbits(64))
provide(RandomUUID())

class RandomUUID4(LibFcn):
    name = prefix + "uuid4"
    sig = Sig([], P.String())
    errcodeBase = 34110
    def __call__(self, state, scope, pos, paramTypes):
        return "{0:08x}-{1:04x}-4{2:03x}-8{3:03x}-{4:016x}".format(state.rand.getrandbits(32), state.rand.getrandbits(16), state.rand.getrandbits(12), state.rand.getrandbits(12), state.rand.getrandbits(64))
provide(RandomUUID4())

########################################################### common probability distributions

class Gaussian(LibFcn):
    name = prefix + "gaussian"
    sig = Sig([{"mu": P.Double()}, {"sigma": P.Double()}], P.Double())
    errcodeBase = 34120
    def __call__(self, state, scope, pos, paramTypes, mu, sigma):
        return state.rand.gauss(mu, sigma)
provide(Gaussian())

